From f9fe263f02cf4f599699a60c592fd3cce3b2bc9a Mon Sep 17 00:00:00 2001
From: Valtteri Vuorikoski <vuori@notcom.org>
Date: Wed, 22 Apr 2026 22:42:27 +0300
Subject: [PATCH] commands/pgp: look for keyid in hashed signature subpackets
 too.

Until now, PGP signature parsing code has looked for the keyid of the
signer ("issuer" in OpenPGP terms) in the unhashed subpackets section of the
signature packet. However this subpacket can also be located in the hashed
subpackets section. This commit makes the lookup search for the keyid in the
hashed section first, and fall back to the unhashed section if needed.

Sequoia-PGP places the subpacket in the hashed section, while GnuPG places it in
the unhashed section. Hence .sig files generated by Sequoia-PGP were previously
unusable by Grub.

Prior code allowed a hashed subpackets section of any size in the .sig file, but
it is now limited to 4k (same as the unhashed section). This should not be an
issue in practice as the typical signature is around 300-400 bytes total.

Signed-off-by: Valtteri Vuorikoski <vuori@notcom.org>
Reviewed-by: Leo Sandoval <lsandova@redhat.com>
Reviewed-by: Andrew Hamilton <adhamilt@gmail.com>
Part-of: <https://gitlab.freedesktop.org/gnu-grub/grub/-/merge_requests/108>

Index: f9fe263f/grub-core/commands/pgp.c
===================================================================
--- f9fe263f.orig/grub-core/commands/pgp.c
+++ f9fe263f/grub-core/commands/pgp.c
@@ -514,6 +514,49 @@ grub_pubkey_write (void *ctxt_, void *bu
 }
 
 static grub_err_t
+parse_signature_keyid (grub_uint64_t *keyid,
+                       const grub_uint8_t *readbuf,
+                       const grub_ssize_t len)
+{
+  const grub_uint8_t *ptr;
+  grub_uint32_t l;
+
+  for (ptr = readbuf; ptr < readbuf + len; ptr += l)
+    {
+      if (*ptr < 192)
+        l = *ptr++;
+      else if (*ptr < 255)
+        {
+          if (ptr + 1 >= readbuf + len)
+            return GRUB_ERR_EOF;
+          l = (((ptr[0] & ~192) << GRUB_CHAR_BIT) | ptr[1]) + 192;
+          ptr += 2;
+        }
+      else
+        {
+          if (ptr + 5 >= readbuf + len)
+            return GRUB_ERR_EOF;
+          l = grub_be_to_cpu32 (grub_get_unaligned32 (ptr + 1));
+          ptr += 5;
+        }
+
+      if (l == 0)
+        return GRUB_ERR_OUT_OF_RANGE;
+      /* will need l bytes in readbuf (l includes 1 byte subpkt type) */
+      if (l > ((readbuf + len) - ptr))
+        return GRUB_ERR_EOF;
+
+      if (*ptr == 0x10 && l >= 8) /* subpkt 16: issuer key ID */
+        {
+          *keyid = grub_get_unaligned64 (ptr + 1);
+          return GRUB_ERR_NONE;
+        }
+    }
+
+  return GRUB_ERR_EOF;
+}
+
+static grub_err_t
 grub_verify_signature_real (struct grub_pubkey_context *ctxt,
 			    struct grub_public_key *pkey)
 {
@@ -530,6 +573,7 @@ grub_verify_signature_real (struct grub_
   grub_uint8_t hash_start[2];
   gcry_mpi_t hmpi;
   grub_uint64_t keyid = 0;
+  grub_err_t keyid_res;
   struct grub_public_subkey *sk;
 
   readbuf = grub_malloc (READBUF_SIZE);
@@ -538,55 +582,37 @@ grub_verify_signature_real (struct grub_
 
   ctxt->hash->write (ctxt->hash_context, &ctxt->v, sizeof (ctxt->v));
   ctxt->hash->write (ctxt->hash_context, &ctxt->v4, sizeof (ctxt->v4));
-  while (rem)
-    {
-      r = grub_file_read (ctxt->sig, readbuf,
-			  rem < READBUF_SIZE ? rem : READBUF_SIZE);
-      if (r < 0)
-	goto fail;
-      if (r == 0)
-	break;
-      ctxt->hash->write (ctxt->hash_context, readbuf, r);
-      rem -= r;
-    }
+
+  if (rem > READBUF_SIZE)
+    goto fail;
+  r = grub_file_read (ctxt->sig, readbuf, rem);
+  if (r != rem)
+    goto fail;
+  ctxt->hash->write (ctxt->hash_context, readbuf, r);
+  /* keyid packet may be in a hashed or unhashed subpacket; check hashed */
+  keyid_res = parse_signature_keyid (&keyid, readbuf, r);
+
   ctxt->hash->write (ctxt->hash_context, &ctxt->v, sizeof (ctxt->v));
   s = 0xff;
   ctxt->hash->write (ctxt->hash_context, &s, sizeof (s));
   ctxt->hash->write (ctxt->hash_context, &headlen, sizeof (headlen));
+
   r = grub_file_read (ctxt->sig, &unhashed_sub, sizeof (unhashed_sub));
   if (r != sizeof (unhashed_sub))
     goto fail;
-  {
-    grub_uint8_t *ptr;
-    grub_uint32_t l;
-    rem = grub_be_to_cpu16 (unhashed_sub);
-    if (rem > READBUF_SIZE)
-      goto fail;
-    r = grub_file_read (ctxt->sig, readbuf, rem);
-    if (r != rem)
-      goto fail;
-    for (ptr = readbuf; ptr < readbuf + rem; ptr += l)
-      {
-	if (*ptr < 192)
-	  l = *ptr++;
-	else if (*ptr < 255)
-	  {
-	    if (ptr + 1 >= readbuf + rem)
-	      break;
-	    l = (((ptr[0] & ~192) << GRUB_CHAR_BIT) | ptr[1]) + 192;
-	    ptr += 2;
-	  }
-	else
-	  {
-	    if (ptr + 5 >= readbuf + rem)
-	      break;
-	    l = grub_be_to_cpu32 (grub_get_unaligned32 (ptr + 1));
-	    ptr += 5;
-	  }
-	if (*ptr == 0x10 && l >= 8)
-	  keyid = grub_get_unaligned64 (ptr + 1);
-      }
-  }
+  rem = grub_be_to_cpu16 (unhashed_sub);
+  if (rem > READBUF_SIZE)
+    goto fail;
+  r = grub_file_read (ctxt->sig, readbuf, rem);
+  if (r != rem)
+    goto fail;
+  if (keyid_res != GRUB_ERR_NONE)
+    {
+      /* keyid was not in hashed packets; check unhashed */
+      keyid_res = parse_signature_keyid (&keyid, readbuf, r);
+      if (keyid_res != GRUB_ERR_NONE)
+        goto fail;
+    }
 
   ctxt->hash->final (ctxt->hash_context);
 
